src/pac.c

```diff
@@ -105,47 +108,65 @@ pac_may_have_muzzy(pac_t *pac) {
 }
 
 static edata_t *
-pac_alloc_real(tsdn_t *tsdn, pac_t *pac, ehooks_t *ehooks, size_t size,
-    size_t alignment, bool zero, bool guarded) {
-	assert(!guarded || alignment <= PAGE);
+pac_alloc_unguarded(tsdn_t *tsdn, pac_t *pac, size_t size, size_t alignment,
+    bool zero, bool *deferred_work_generated) {
+	ehooks_t *ehooks = pac_ehooks_get(pac);
 
 	edata_t *edata = ecache_alloc(tsdn, pac, ehooks, &pac->ecache_dirty,
-	    NULL, size, alignment, zero, guarded);
+	    NULL, size, alignment, zero, false);
 
 	if (edata == NULL && pac_may_have_muzzy(pac)) {
 		edata = ecache_alloc(tsdn, pac, ehooks, &pac->ecache_muzzy,
-		    NULL, size, alignment, zero, guarded);
+		    NULL, size, alignment, zero, false);
 	}
 	if (edata == NULL) {
 		edata = ecache_alloc_grow(tsdn, pac, ehooks,
-		    &pac->ecache_retained, NULL, size, alignment, zero,
-		    guarded);
+		    &pac->ecache_retained, NULL, size, alignment, zero);
 		if (config_stats && edata != NULL) {
 			atomic_fetch_add_zu(&pac->stats->pac_mapped, size,
 			    ATOMIC_RELAXED);
 		}
 	}
 
 	return edata;
+
 }
 
 static edata_t *
-pac_alloc_new_guarded(tsdn_t *tsdn, pac_t *pac, ehooks_t *ehooks, size_t size,
-    size_t alignment, bool zero) {
-	assert(alignment <= PAGE);
-
-	size_t size_with_guards = size + SAN_PAGE_GUARDS_SIZE;
-	/* Alloc a non-guarded extent first.*/
-	edata_t *edata = pac_alloc_real(tsdn, pac, ehooks, size_with_guards,
-	    /* alignment */ PAGE, zero, /* guarded */ false);
-	if (edata != NULL) {
-		/* Add guards around it. */
-		assert(edata_size_get(edata) == size_with_guards);
-		san_guard_pages(tsdn, ehooks, edata, pac->emap, true, true,
-		    true);
-	}
-	assert(edata == NULL || (edata_guarded_get(edata) &&
-	    edata_size_get(edata) == size));
+pac_alloc_guarded(tsdn_t *tsdn, pac_t *pac, size_t size, size_t alignment,
+    bool zero, bool *deferred_work_generated) {
+	ehooks_t *ehooks = pac_ehooks_get(pac);
+
+	edata_t *edata;
+	edata = ecache_alloc(tsdn, pac, ehooks, &pac->ecache_dirty, NULL, size,
+	    alignment, zero, true);
+
+	if (edata == NULL && pac_may_have_muzzy(pac)) {
+		edata = ecache_alloc(tsdn, pac, ehooks, &pac->ecache_muzzy,
+		    NULL, size, alignment, zero, true);
+	}
+	if (edata == NULL) {
+		edata = ecache_alloc(tsdn, pac, ehooks, &pac->ecache_retained,
+		    NULL, size, alignment, zero, true);
+
+		if (edata != NULL && config_prof) {
+			extent_gdump_add(tsdn, edata);
+		}
+	}
```

Comment:
Discussed offline. Totally agree, that `dirty->muzzy->retained` duplication sucks and must be avoided. I just went to redo it and remembered the original reasoning behind these proposed design changes.

The original approach dragged `guarded` flag deep into `extent.c` with some inevitable assumptions about guarded control flow upstream. For instance, `ecache_alloc_grow` won't actually grow for guarded, because it knows it was called solely to perform retained lookup, without growing intentions. This, while involves less duplication, makes it very implicit for the uninitiated: "Why doesn't `ecache_alloc_grow` grow retained for guarded allocations?". To answer this question, readers should go up the stack to PAC and find the answer there.

My approach, on the other hand, tired to localize most changes in the control flow part of this - PAC. But yes, that came at the cost of duplication, which I too don't want to pay.